Prevalidation of B1 upgrades

oyvindronningstad opened the pull request:
- Refactor b0 validation code
 - Create ABI into b0 validation code

oyvindronningstad commented:
cc @sigvartmh @hakonfam @osaether

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:45: WARNING:LINE_SPACING: Missing a blank line after declarations
#45: FILE: include/bl_validation.h:64:
+	const u32_t validation_info_magic[] = {VALIDATION_INFO_MAGIC};
+	if (memeq(vinfo->magic, validation_info_magic, CONFIG_FW_MAGIC_LEN)) {

-:109: WARNING:BLOCK_COMMENT_STYLE: Block comments use a trailing */ on a separate line
#109: FILE: include/fw_metadata.h:190:
+ * */

-:126: WARNING:COMPLEX_MACRO: Macros with complex values should be enclosed in parentheses
#126: FILE: include/fw_metadata.h:208:
+#define FW_INFO_OFFSETS FW_INFO_OFFSET0, FW_INFO_OFFSET1, FW_INFO_OFFSET2

-:128: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#128: FILE: include/fw_metadata.h:210:
+#if (CONFIG_FW_FIRMWARE_INFO_OFFSET != FW_INFO_OFFSET0 \

-:129: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#129: FILE: include/fw_metadata.h:211:
+  && CONFIG_FW_FIRMWARE_INFO_OFFSET != FW_INFO_OFFSET1 \$

-:130: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#130: FILE: include/fw_metadata.h:212:
+  && CONFIG_FW_FIRMWARE_INFO_OFFSET != FW_INFO_OFFSET2) \$

-:131: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#131: FILE: include/fw_metadata.h:213:
+ || FW_INFO_OFFSET_COUNT != 3$

-:141: WARNING:LONG_LINE: line over 80 characters
#141: FILE: include/fw_metadata.h:223:
+	static_assert(sizeof(offsets)/sizeof(offsets[0]) == FW_INFO_OFFSET_COUNT,

-:141: WARNING:LINE_SPACING: Missing a blank line after declarations
#141: FILE: include/fw_metadata.h:223:
+	u32_t offsets[] = {FW_INFO_OFFSETS};
+	static_assert(sizeof(offsets)/sizeof(offsets[0]) == FW_INFO_OFFSET_COUNT,

-:141: WARNING:ARRAY_SIZE: Prefer ARRAY_SIZE(offsets)
#141: FILE: include/fw_metadata.h:223:
+	static_assert(sizeof(offsets)/sizeof(offsets[0]) == FW_INFO_OFFSET_COUNT,

-:346: WARNING:LINE_SPACING: Missing a blank line after declarations
#346: FILE: samples/bootloader/src/main.c:93:
+	u32_t *vector_table = (u32_t *)fw_info->firmware_address;
+	VTOR = fw_info->firmware_address;

-:440: WARNING:LINE_SPACING: Missing a blank line after declarations
#440: FILE: samples/nrf9160/secure_services/src/main.c:63:
+	bool valid = spm_prevalidate_b1_upgrade(0x8000, 0x8000);
+	printk("\nS0 valid? %s\n", valid ? "True" : "False");

-:588: ERROR:INITIALISED_STATIC: do not initialise statics to NULL
#588: FILE: subsys/bootloader/bl_client/bl_client.c:22:
+	static const struct fw_abi_info *abi = NULL;

-:589: WARNING:LINE_SPACING: Missing a blank line after declarations
#589: FILE: subsys/bootloader/bl_client/bl_client.c:23:
+	static const struct fw_abi_info *abi = NULL;
+	if (!abi) {

-:599: ERROR:INITIALISED_STATIC: do not initialise statics to NULL
#599: FILE: subsys/bootloader/bl_client/bl_client.c:33:
+	static const struct fw_abi_info *abi = NULL;

-:600: WARNING:LINE_SPACING: Missing a blank line after declarations
#600: FILE: subsys/bootloader/bl_client/bl_client.c:34:
+	static const struct fw_abi_info *abi = NULL;
+	if (!abi) {

-:610: ERROR:INITIALISED_STATIC: do not initialise statics to NULL
#610: FILE: subsys/bootloader/bl_client/bl_client.c:44:
+	static const struct fw_abi_info *abi = NULL;

-:611: WARNING:LINE_SPACING: Missing a blank line after declarations
#611: FILE: subsys/bootloader/bl_client/bl_client.c:45:
+	static const struct fw_abi_info *abi = NULL;
+	if (!abi) {

-:621: ERROR:INITIALISED_STATIC: do not initialise statics to NULL
#621: FILE: subsys/bootloader/bl_client/bl_client.c:55:
+	static const struct fw_abi_info *abi = NULL;

-:622: WARNING:LINE_SPACING: Missing a blank line after declarations
#622: FILE: subsys/bootloader/bl_client/bl_client.c:56:
+	static const struct fw_abi_info *abi = NULL;
+	if (!abi) {

-:623: WARNING:LONG_LINE: line over 80 characters
#623: FILE: subsys/bootloader/bl_client/bl_client.c:57:
+		abi = fw_abi_find(BL_VALIDATE_FW_ABI_ID, BL_VALIDATE_FW_ABI_FLAGS,

-:658: WARNING:LONG_LINE: line over 80 characters
#658: FILE: subsys/bootloader/bl_client/bl_client.c:92:
+	return get_bl_sha256_abi()->abi.bl_sha256_verify(data, data_len, expected);

-:661: WARNING:LONG_LINE: line over 80 characters
#661: FILE: subsys/bootloader/bl_client/bl_client.c:95:
+int bl_secp256r1_validate(const u8_t *hash, u32_t hash_len, const u8_t *public_key, const u8_t *signature)

-:663: WARNING:LONG_LINE: line over 80 characters
#663: FILE: subsys/bootloader/bl_client/bl_client.c:97:
+	return get_bl_secp256r1_abi()->abi.bl_secp256r1_validate(hash, hash_len, public_key, signature);

-:669: WARNING:MULTILINE_DEREFERENCE: Avoid multiple line dereference - prefer 'abi.bl_validate_firmware'
#669: FILE: subsys/bootloader/bl_client/bl_client.c:103:
+	return get_bl_validate_firmware_abi()->abi.
+		bl_validate_firmware(fw_dst_address, fw_src_address);

-:711: WARNING:LINE_SPACING: Missing a blank line after declarations
#711: FILE: subsys/bootloader/bl_crypto/bl_crypto_cc310_hash.c:47:
+	size_t len_words = ROUND_UP(n, 4) / 4;
+	for (size_t i = 0; i < len_words; i++) {

-:878: ERROR:OPEN_BRACE: that open brace { should be on the previous line
#878: FILE: subsys/bootloader/bl_validation/bl_validation.c:40:
+	if (!(((u32_t)fw_info >= fw_src_address)
+		&& (((u32_t)fw_info + sizeof(*fw_info))
+			< (fw_src_address + fw_info->firmware_size))))
+	{

-:958: WARNING:LONG_LINE: line over 80 characters
#958: FILE: subsys/bootloader/bl_validation/bl_validation.c:120:
+				BL_VALIDATE_FW_ABI_VER, sizeof(struct bl_validate_fw_abi)),

- total: 5 errors, 23 warnings, 916 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.



```
## Gitlint issues

Commit 3c27c1d625:
1: UC2 Body does not contain a 'Signed-off-by:' line

Commit 52bd4a6820:
3: UC4 Line exceeds max length (74>72): "Since it needs to be protected, this works better in both nrf91 and nrf5x."

Commit fb9a0a128d:
1: UC2 Body does not contain a 'Signed-off-by:' line



## Identity/Emails issues

3c27c1d62518dd989ed1686f1280b8ef84454a9d: author email (Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>) needs to match one of the signed-off-by entries.

fb9a0a128d367322cc012cea9b19920d6e4738c1: author email (Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>) needs to match one of the signed-off-by entries.

hakonfam commented on include/fw_metadata.h:
```diff
@@ -199,6 +201,36 @@ fw_firmware_info_get(u32_t firmware_address)
 	return NULL;
 }
 
+#define FW_INFO_OFFSET0 0x200
```
This looks weird, what is this?

oyvindronningstad replied:
Yep, will do. BTW I'd appreciate ideas here, like which offsets should be allowed, and how to implement it differently etc.

oyvindronningstad replied:
Added comment

oyvindronningstad commented on include/fw_metadata.h:
```diff
@@ -199,6 +201,36 @@ fw_firmware_info_get(u32_t firmware_address)
 	return NULL;
 }
 
+#define FW_INFO_OFFSET0 0x200
```
Vector tables are always the same size on the same hardware. From what I've seen, we can't go much smaller, but we could look more into it.

oyvindronningstad replied:
On nrf91, the vector table seems to take 0x40 in total, but the total space used in front of the fw_info is 0x144 in a lot of cases (maybe IRQ handlers etc?) I think 0x200 is reasonable, but it could be squeezed a bit. What do you think?

oyvindronningstad commented on samples/bootloader/pm.yml:
```diff
@@ -1,9 +1,12 @@
   placement: {after: start}
 
+b0:
+  span: [b0_image, provision]
```
Are we already using OTP for provision data?

oyvindronningstad replied:
As we found out, OTP was used, but that was reverted for another change. Will keep it as is for now.

oyvindronningstad commented on subsys/spm/secure_services.c:
```diff
@@ -111,3 +112,13 @@ int spm_request_random_number(u8_t *output, size_t len, size_t *olen)
+__TZ_NONSECURE_ENTRY_FUNC
+bool spm_prevalidate_b1_upgrade(u32_t dst_addr, u32_t src_addr)
+{
+	bool result = bl_validate_firmware(dst_addr, src_addr);
```
This is split for easier placement of breakpoints.

hakonfam replied:
ah, since this is a magic limit which the debugger won't be able to follow. Get it.

oyvindronningstad commented on include/fw_metadata.h:
```diff
@@ -184,11 +184,13 @@ static inline bool memeq(const void *expected, const void *actual, u32_t len)
+ * */
 static inline const struct fw_firmware_info *
-fw_firmware_info_get(u32_t firmware_address)
+fw_firmware_info_check(u32_t finfo_addr)
```
I'll go through the commits again after I'm back.

oyvindronningstad replied:
Rewrote some commit messages

oyvindronningstad commented on include/bl_validation.h:
```diff
@@ -57,24 +58,51 @@ OFFSET_CHECK(struct fw_validation_pointer, validation_info,
 
-#endif /* BOOTLOADER_H__ */
+
+#define BL_VALIDATE_FW_ABI_ID 0x1101
```
Made a Kconfig template for this and used it for bl_crypto ABIs as well.

oyvindronningstad commented:
@hakonfam All your comments should be addressed now.

oyvindronningstad commented:
Moved some changes into their own PRs: #1183, #1184, #1185

hakonfam commented on include/fw_metadata.h:
```diff
@@ -200,6 +203,42 @@ fw_firmware_info_get(u32_t firmware_address)
+static_assert(ARRAY_SIZE(allowed_offsets) == FW_INFO_OFFSET_COUNT,
+		"Mismatch in the number of allowed offsets.");
+
+#if (FW_INFO_OFFSET_COUNT != 3) || \
```
Why do we need this check when we have the static assert above?

oyvindronningstad replied:
If offsets are added to the list, but this #if is not updated, this check will catch that.

hakonfam commented on include/fw_metadata.h:
```diff
@@ -200,6 +203,42 @@ fw_firmware_info_get(u32_t firmware_address)
+
+/* Search for the firmware_info structure inside the firmware. */
+static inline const struct fw_firmware_info *
+fw_firmware_info_find(u32_t firmware_address)
```
Should we address this whole "fw_firmware" issue while we are at it? It is starting to look very silly :D

oyvindronningstad replied:
Haha, gladly. Suggestions are very welcome.

hakonfam replied:
So, "B0", "B1" and so on is conseptual names for the different stages in the boot chain. We can not use these for anything. 

We need some identifier for OUR B0, that is what we now call "Secure Boot" or "Bootloader", 
IMO we should create a PR cleaning this up once and for all. My proposal is NSB (no pun intended) for Nordic Secure Boot.  That way this would be `nsb_metadata.h`, `nsb_fw_info_find` and so on.

oyvindronningstad replied:
Note that fw_metadata is its own subsys. This was done because it might eventually be used other places as well. But I definitely support the proposal to rename the `bootloader` subsys to something more specific :).

sigvartmh replied:
at least every static function in the library does not need to be namespaced/prefixed to `fw_`

sigvartmh commented on include/fw_metadata.h:
```diff
@@ -184,11 +185,13 @@ static inline bool memeq(const void *expected, const void *actual, u32_t len)
+ */
 static inline const struct fw_firmware_info *
-fw_firmware_info_get(u32_t firmware_address)
+fw_firmware_info_check(u32_t finfo_addr)
```
Just a suggestion for variable name feel free to ignore
```suggestion
fw_firmware_info_check(u32_t fw_info_addr)

sigvartmh commented on include/fw_metadata.h:
```diff
@@ -200,6 +203,42 @@ fw_firmware_info_get(u32_t firmware_address)
+static const u32_t allowed_offsets[] = {FW_INFO_OFFSET0, FW_INFO_OFFSET1,
+					FW_INFO_OFFSET2};
+
+static_assert(ARRAY_SIZE(allowed_offsets) == FW_INFO_OFFSET_COUNT,
```
Maybe more zephyr-ish does, however, it makes it less portable. Feel free to ignore
```suggestion
BUILD_ASSERT_MSG(ARRAY_SIZE(allowed_offsets) == FW_INFO_OFFSET_COUNT,
"Mismatch in the number of allowed offsets.");

sigvartmh commented on samples/bootloader/src/main.c:
```diff
@@ -182,8 +182,8 @@ void main(void)
 	u32_t s1_addr = s1_address_read();
-	const struct fw_firmware_info *s0_info = fw_firmware_info_get(s0_addr);
-	const struct fw_firmware_info *s1_info = fw_firmware_info_get(s1_addr);
+	const struct fw_firmware_info *s0_info = fw_firmware_info_find(s0_addr);
```
```suggestion
	const struct fw_firmware_info *s0_info = fw_find_firmware_info((s0_addr);
```
or

```suggestion
	const struct fw_firmware_info *s0_info = fw_find_metadata((s0_addr);

sigvartmh commented on subsys/fw_metadata/Kconfig:
```diff
@@ -18,11 +18,13 @@ config FW_VALIDATION_METADATA_OFFSET
+	  The location of firmware info inside this firmware. Valid values are
+	  0x200, 0x400 and 0x800. Compatible readers of FW info should search
+	  all the possible offsets. Note that all space between the vector table
+	  and this address will be unused.
```
Maybe change from future tense to present tense
```suggestion	  
	  and this address is unused.

sigvartmh commented on subsys/fw_metadata/Kconfig:
```diff
@@ -18,11 +18,13 @@ config FW_VALIDATION_METADATA_OFFSET
 config FW_FIRMWARE_INFO_OFFSET
-	hex "The location of firmware info inside firmware."
-	default 0x800
+	hex "The location of firmware info inside this firmware."
```
Would it ever be unclear what this firmware is related to incase multiple systems use it?

oyvindronningstad replied:
Changed to "inside the current firmware image."

sigvartmh commented on subsys/fw_metadata/fw_metadata.c:
```diff
@@ -51,7 +51,8 @@ __fw_info struct fw_firmware_info m_firmware_info =
 void fw_abi_provide(u32_t address)
 {
-	const struct fw_firmware_info *their_fw_info = fw_firmware_info_get(address);
+	const struct fw_firmware_info *their_fw_info
```
Is their maybe a little vague? I'm not really sure if I have a better suggestion.

provided_fw_info?

sigvartmh commented on subsys/bootloader/Kconfig:
```diff
@@ -12,13 +12,15 @@ config MCUBOOT_BUILD_S1_VARIANT
-	  which gives access to the cryptographic functions
-	  from the secure bootloader. This will only work
-	  when the secure bootloader is enabled.
+	  Set this option to enable the bootloader client API which gives access
```
```suggestion
	  Set this option to enable the bootloader client API, which gives access

sigvartmh commented on subsys/bootloader/Kconfig:
```diff
@@ -12,13 +12,15 @@ config MCUBOOT_BUILD_S1_VARIANT
-	  when the secure bootloader is enabled.
+	  Set this option to enable the bootloader client API which gives access
+	  to validation functions as well as cryptographic primitives from the
+	  secure bootloader. This will only work when the secure bootloader is
```
```suggestion
	  secure bootloader. This only works when the secure bootloader is

sigvartmh commented on samples/bootloader/src/main.c:
```diff
@@ -104,11 +101,15 @@ extern u32_t _vector_table_pointer;
 
-	if (!verify_firmware((u32_t)address)) {
+	if (!verify_firmware(fw_info->firmware_address)) {
+		printk("Failed to validate!\n\r");
```
Should we try to mark failures as <err> and <inf> etc like zephyr? Could we re-use some print macros maybe?
```suggestion
		printk("err: Failed to validate!\n\r");

oyvindronningstad replied:
I can't find that this is a convention. Maybe "err:" is added automatically by the logging subsys, but we are using printk instead.

sigvartmh commented on samples/bootloader/src/main.c:
```diff
@@ -148,25 +149,33 @@ static void boot_from(u32_t *address)
+	__ISB(); /* Flush and refill pipeline with updated permissions */
 
-	VTOR = (u32_t)address;
+	VTOR = fw_info->firmware_address;
```
could be that I'm misunderstanding something but couldn't this be possible?
```suggestion
	VTOR = vector_table

oyvindronningstad replied:
That's correct, but `fw_info->firmware_address` will read it directly from flash, so I opted for that.

sigvartmh commented on samples/bootloader/src/main.c:
```diff
@@ -108,7 +47,8 @@ static void boot_from(const struct fw_firmware_info *fw_info)
 		fw_info->firmware_address);
 
-	if (!verify_firmware(fw_info->firmware_address)) {
+	if (!bl_validate_firmware_local(fw_info->firmware_address,
```
what does local refer to here? That it's not used as an external abi?

oyvindronningstad replied:
Correct. Naming suggestions welcome :)

sigvartmh commented on subsys/bootloader/bl_validation/bl_validation.c:
```diff
@@ -0,0 +1,203 @@
+
+
+/* Static asserts to ensure compatibility */
+OFFSET_CHECK(struct fw_validation_info, magic, 0);
```
BUILD_ASSERT?

hakonfam commented:
I still feel that this PR is too big for proper review, and it looks like it can still be quite easily split into multiple PRs. How do you feel about splitting this into more PRs @oyvindronningstad ?

oyvindronningstad commented:
> I still feel that this PR is too big for proper review, and it looks like it can still be quite easily split into multiple PRs. How do you feel about splitting this into more PRs @oyvindronningstad ?

Moved 2 more commits into separate PRs: #1218 #1219

oyvindronningstad commented:
> > I still feel that this PR is too big for proper review, and it looks like it can still be quite easily split into multiple PRs. How do you feel about splitting this into more PRs @oyvindronningstad ?
> 
> Moved 2 more commits into separate PRs: #1218 #1219

And removed client changes. They will be PRed after this one.

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## Kconfig issues

```
Undefined Kconfig symbols:

warning: undefined symbol SECURE_BOOT_CLIENT_API:

- Referenced at ../nrf/subsys/fw_metadata/Kconfig.template.fw_metadata_abi:2:

config BL_ROT_VERIFY_ABI_FLAGS
	int
	default 2
	depends on (SECURE_BOOT_CRYPTO || SECURE_BOOT_CLIENT_API) && (IS_SECURE_BOOTLOADER || SECURE_BOOT)
	help
	  Flags for the $(ABI) ABI.

- Referenced at ../nrf/subsys/fw_metadata/Kconfig.template.fw_metadata_abi:8:

config BL_ROT_VERIFY_ABI_VER
	int
	default 1
	depends on (SECURE_BOOT_CRYPTO || SECURE_BOOT_CLIENT_API) && (IS_SECURE_BOOTLOADER || SECURE_BOOT)
	help
	  The current or requested version for the $(ABI) ABI.
	  This config is used by both the provider and client of the ABI.

- Referenced at ../nrf/subsys/fw_metadata/Kconfig.template.fw_metadata_abi:15:

config BL_ROT_VERIFY_ABI_MAX_VER
	int
	default 255
	depends on (SECURE_BOOT_CRYPTO || SECURE_BOOT_CLIENT_API) && (IS_SECURE_BOOTLOADER || SECURE_BOOT)
	help
	  The maximum requested version for the $(ABI) ABI.

- Referenced at ../nrf/subsys/fw_metadata/Kconfig.template.fw_metadata_abi:2:

config BL_SHA256_ABI_FLAGS
	int
	default 0
	depends on (SECURE_BOOT_CRYPTO || SECURE_BOOT_CLIENT_API) && (IS_SECURE_BOOTLOADER || SECURE_BOOT)
	help
	  Flags for the $(ABI) ABI.

- Referenced at ../nrf/subsys/fw_metadata/Kconfig.template.fw_metadata_abi:8:

config BL_SHA256_ABI_VER
	int
	default 1
	depends on (SECURE_BOOT_CRYPTO || SECURE_BOOT_CLIENT_API) && (IS_SECURE_BOOTLOADER || SECURE_BOOT)
	help
	  The current or requested version for the $(ABI) ABI.
	  This config is used by both the provider and client of the ABI.

- Referenced at ../nrf/subsys/fw_metadata/Kconfig.template.fw_metadata_abi:15:

config BL_SHA256_ABI_MAX_VER
	int
	default 255
	depends on (SECURE_BOOT_CRYPTO || SECURE_BOOT_CLIENT_API) && (IS_SECURE_BOOTLOADER || SECURE_BOOT)
	help
	  The maximum requested version for the $(ABI) ABI.

- Referenced at ../nrf/subsys/fw_metadata/Kconfig.template.fw_metadata_abi:2:

config BL_SECP256R1_ABI_FLAGS
	int
	default 1
	depends on (SECURE_BOOT_CRYPTO || SECURE_BOOT_CLIENT_API) && (IS_SECURE_BOOTLOADER || SECURE_BOOT)
	help
	  Flags for the $(ABI) ABI.

- Referenced at ../nrf/subsys/fw_metadata/Kconfig.template.fw_metadata_abi:8:

config BL_SECP256R1_ABI_VER
	int
	default 1
	depends on (SECURE_BOOT_CRYPTO || SECURE_BOOT_CLIENT_API) && (IS_SECURE_BOOTLOADER || SECURE_BOOT)
	help
	  The current or requested version for the $(ABI) ABI.
	  This config is used by both the provider and client of the ABI.

- Referenced at ../nrf/subsys/fw_metadata/Kconfig.template.fw_metadata_abi:15:

config BL_SECP256R1_ABI_MAX_VER
	int
	default 255
	depends on (SECURE_BOOT_CRYPTO || SECURE_BOOT_CLIENT_API) && (IS_SECURE_BOOTLOADER || SECURE_BOOT)
	help
	  The maximum requested version for the $(ABI) ABI.

- Referenced at ../nrf/subsys/fw_metadata/Kconfig.template.fw_metadata_abi:2:

config BL_VALIDATE_FW_ABI_FLAGS
	int
	default 0
	depends on (SECURE_BOOT_VALIDATION || SECURE_BOOT_CLIENT_API) && (IS_SECURE_BOOTLOADER || SECURE_BOOT)
	help
	  Flags for the $(ABI) ABI.

- Referenced at ../nrf/subsys/fw_metadata/Kconfig.template.fw_metadata_abi:8:

config BL_VALIDATE_FW_ABI_VER
	int
	default 1
	depends on (SECURE_BOOT_VALIDATION || SECURE_BOOT_CLIENT_API) && (IS_SECURE_BOOTLOADER || SECURE_BOOT)
	help
	  The current or requested version for the $(ABI) ABI.
	  This config is used by both the provider and client of the ABI.

- Referenced at ../nrf/subsys/fw_metadata/Kconfig.template.fw_metadata_abi:15:

config BL_VALIDATE_FW_ABI_MAX_VER
	int
	default 255
	depends on (SECURE_BOOT_VALIDATION || SECURE_BOOT_CLIENT_API) && (IS_SECURE_BOOTLOADER || SECURE_BOOT)
	help
	  The maximum requested version for the $(ABI) ABI.
```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

oyvindronningstad commented:
@sigvartmh your comments are addressed either here, in #1219, or on the branch that will be PRed later.

hakonfam commented:
Great job splitting out to several PRs.

I still, however, feel that this should be more than a single commit.

Reading the current commit message gives me the impression that it could have been split into:

(1) `bootloader: Move validation logic into bl_validation from fw_metadata and bootloader.`

And (2)`expose validation through ABI`.
(3)`Create a Kconfig template for ABI defines.`
(4)`Refactor validation logic and bl_crypto's ABI defines.`

(5)`Some configs are removed altogether because they were duplicated between
the bootloader and fw_metadata. E.g. FW_SHA256 etc.`

5 commits. Maybe not all of these are trivial to split up without much work etc. in which case I'm fine. But there should be a good reason not to split this up more IMO.

oyvindronningstad commented:
> 5 commits. Maybe not all of these are trivial to split up without much work etc. in which case I'm fine. But there should be a good reason not to split this up more IMO.

Do you want separate PRs or just separate commits? I'll see what I can do 👍

hakonfam commented:
> > 5 commits. Maybe not all of these are trivial to split up without much work etc. in which case I'm fine. But there should be a good reason not to split this up more IMO.
> 
> Do you want separate PRs or just separate commits? I'll see what I can do 👍

Separate commits would suffice I guess, it will make it easier to review at least.
IMO Separate PRs is always the best solution when the PR is sane.
I guess its a consideration between effort and gain for you to decide :)

oyvindronningstad commented:
> > > 5 commits. Maybe not all of these are trivial to split up without much work etc. in which case I'm fine. But there should be a good reason not to split this up more IMO.
> > 
> > 
> > Do you want separate PRs or just separate commits? I'll see what I can do 👍
> 
> Separate commits would suffice I guess, it will make it easier to review at least.
> IMO Separate PRs is always the best solution when the PR is sane.
> I guess its a consideration between effort and gain for you to decide :)

Split into 4 commits, one of which I moved to #1223.

hakonfam commented on subsys/bootloader/bl_validation/bl_validation.c:
```diff
@@ -0,0 +1,75 @@
+	fw_info = fw_find_firmware_info(address);
+
+	if (!fw_info) {
+		printk("Could not find valid firmware info inside "
```
Should we use a logger in this file instead of `printk`, as this is now a subsystem and not part of a the sample code?

oyvindronningstad replied:
It seems to add about 10k to the b0 size, which is a bit inconvenient, so I propose we leave it for now.

hakonfam replied:
IMO size with logging is not something we should worry about unless it causes issues with building (e.g. stepping over regions and so on).

oyvindronningstad replied:
It's non-trivial to do, so I'll change it in a later PR.

hakonfam commented on samples/bootloader/src/main.c:
```diff
@@ -24,6 +24,7 @@
 #endif
 #endif
 
+
```
remove

oyvindronningstad replied:
done

oyvindronningstad commented:
> The comment in the last commit message is outdated
> 
> `Move much of the header file content to c ...` as this is done in a different commit now.

Fixed
